File: urldecode/urldecode.js
// ATOMIC FUNCTION: urldecode(s)
// TAGS: string
// LANG: node.js
// FUNCTION: urldecode(s)
function urldecode(s) {
  return decodeURIComponent(s);
}
var start_val = 'mycgi%3Ffoo%3D';
var expected_end_val="mycgi?foo=";
var end_val=urldecode(start_val);

// TEST
if(end_val==expected_end_val){
    console.log('PASS');
}else{
    console.log('FAIL');
}

File: urldecode/urldecode.java

import java.nio.charset.StandardCharsets;

public class urldecode {
    private static String urldecode(String value) {

        return java.net.URLDecoder.decode(value, StandardCharsets.UTF_8);

    }
    public static void main(String args[]) {
        String start_val ="mycgi%3Ffoo%3D";
        String expected_end_val ="mycgi?foo=";
        String end_val = urldecode(start_val);


        // TEST
        if(end_val.equals( expected_end_val) ){
            System.out.println("PASS");
        } else {
            System.out.println("FAIL");
        }
    }

}
